### PR TITLE
Cross-bitness support in JitInterface

### PR DESCRIPTION
**Pull request by echesakovMSFT**
This PR:
1. Removes `TransparentProxyObject` and `RealProxyObject` not used in CoreCLR
2. Skip calling `InlinedCallFrame::GetEEInfo` for R2R compilation
3. Introduce target-specific `OFFSETOF__DelegateObject__target` `OFFSETOF__DelegateObject__methodPtr` `OFFSETOF__DelegateObject__methodPtrAux` `OFFSETOF__PtrArray__m_Array` constants to be used in `CEEInfo::getEEInfo`

@jkotas Do you think `getEEInfo` would be appropriate place to put assertions like `_ASSERTE(OFFSETOF__DelegateObject__target == DelegateObject::GetOffsetOfTarget())` and enable them only in non-crossbitness scenario or there is a better place to do this?

### CODE REVIEW

File: src/vm/jitinterface.cpp [hunk 1]
```diff
@@ -10090,24 +10090,28 @@ void CEEInfo::getEEInfo(CORINFO_EE_INFO *pEEInfoOut)
...
 
-    InlinedCallFrame::GetEEInfo(&pEEInfoOut->inlinedCallFrameInfo);
+    if (!IsReadyToRunCompilation())
+    {
+        // inlinedCallFrameInfo is not used for R2R compilation
+        InlinedCallFrame::GetEEInfo(&pEEInfoOut->inlinedCallFrameInfo);
+    }
 
     // Offsets into the Thread structure
     pEEInfoOut->offsetOfThreadFrame = Thread::GetOffsetOfCurrentFrame();
```
Comment (jkotas) [inline src/vm/jitinterface.cpp]: The Thread fields should be under the `!IsReadyToRunCompilation()` as well.

**Comment by echesakovMSFT** [comment]
@jkotas I have a question. Since `OFFSETOF__PtrArray__m_Array` is defined in amd64/asmconstants.h what do you think is the "right" approach:
1. Remove `OFFSETOF__PtrArray__m_Array` along with `ASMCONSTANTS_C_ASSERT` from amd64/asmconstants.h (which is not safe unless we add such assertions somewhere else)
2. Rename introduced in object.h `OFFSETOF__PtrArray__m_Array` to something like `OFFSETOF__PtrArray__Data` (which would create duplicity)
3. Put `#ifndef OFFSETOF__PtrArray__m_Array`/`#endif` around each definition and keep assertion in amd64/asmconstants.h

File: src/vm/jitinterface.cpp (hunk 1 above)
Comment (echesakovMSFT) [reply src/vm/jitinterface.cpp]: Fixed. Also added zeroing memory for `inlinedCallFrameInfo` in R2R compilation

**Comment by jkotas** [comment]
I would duplicate it for now (your option 2). You can keep the name similar - e.g. just add underscore to the other one, so that it is easy to find.

**Comment by jkotas** [comment]
> Do you think getEEInfo would be appropriate place to put assertions like `_ASSERTE(OFFSETOF__DelegateObject__target == DelegateObject::GetOffsetOfTarget())`

Yes. I cannot think about anything simple that is obviously better.

File: src/vm/jitinterface.cpp [hunk 2]
```diff
@@ -10090,24 +10090,43 @@ void CEEInfo::getEEInfo(CORINFO_EE_INFO *pEEInfoOut)
...
+        ZeroMemory(&pEEInfoOut->inlinedCallFrameInfo, sizeof(pEEInfoOut->inlinedCallFrameInfo));
+
+        pEEInfoOut->offsetOfThreadFrame = (DWORD)-1;
+        pEEInfoOut->offsetOfGCState     = (DWORD)-1;
+    }
 
-    // Offsets into the Thread structure
-    pEEInfoOut->offsetOfThreadFrame = Thread::GetOffsetOfCurrentFrame();
-    pEEInfoOut->offsetOfGCState     = Thread::GetOffsetOfGCFlag();
+#if (!defined(_WIN64) && !defined(_TARGET_64BIT_)) || (defined(_WIN64) && defined(_TARGET_64BIT_))
```
Comment (jkotas) [inline src/vm/jitinterface.cpp]: Would !define(CROSS_COMPILE) work here?

Comment (echesakovMSFT) [reply src/vm/jitinterface.cpp]: My original intent was to run these assertions in every current scenario

Comment (echesakovMSFT) [reply src/vm/jitinterface.cpp]: @jkotas I think I might need such macro somewhere else later. I would define `CROSSBITNESS_COMPILE` in src/inc/crosscomp.h

**Pull request merged**